### ISSUE TITLE
[question] Unexpected "exports_source" behaviour with regard to layout's "folder.root"

### ISSUE THREAD

**Issue by Adnn** [conan-io/conan#14949]
Making a small example for another issue, I tried to use `exports_source` attribute.

I expected the provided paths to be relative to the `folder.root` specified in `layout()`, but they appear to be relative to the recipe folder.

It seemed the [layout documentation](https://docs.conan.io/2/reference/conanfile/methods/layout.html#self-folders) would imply it:

> self.folders.root : Specifies a parent directory where the sources [...] are located specifically when the conanfile.py is located in a separated subdirectory.

In a nutshell, I expected `exports_sources` to look for sources in the `layout()` attribute used to specify sources location.

* Is this as expected?
* If so, how to use `exports_sources` when the conanfile.py is in a subfolder? (since there is an assertion failing when the exported sources start with `..`).

**Comment by memsharded** [comment]
Hi @Adnn 

Thanks for your question.
The ``exports_source`` attribute when the recipe is a subfolder is not really recommended, and instead the ``export_sources()`` method is encouraged. You can see an example here: https://docs.conan.io/2/examples/conanfile/layout/conanfile_in_subfolder.html#examples-conanfile-layout-conanfile-in-subfolder

As you can see in the example, it needs to refer to the parent folder with ``folder = os.path.join(self.recipe_folder, "..")``, because the ``exports_source`` attribute cannot really refer to that parent, even with ``layout()``, because ``layout()`` happens later in the flow, as it requires settings+options to evaluate, but ``export`` happens irrespective of settings and options (you can even run ``conan export`` that doesn't accept any setting, option or profile)

**Comment by Adnn** [comment]
Hi @memsharded

Thank you for this clarification, this completely answers my question.

**Comment by RobertHabrich** [comment]
> since there is an assertion failing when the exported sources start with ..

there should probably be a better error message for this case? I'm trying conan 2.0 right now with our internal scripts that often use relative paths like `exports_sources = "../CMakeLists.txt"`. when trying to run `conan export ./conan --user USER --channel CHANNEL`, I get the following error, which is not very helpful:

```
ERROR: Traceback (most recent call last):
  File "conan\cli\cli.py", line 281, in main
  File "conan\cli\cli.py", line 191, in run
  File "conan\cli\command.py", line 142, in run
  File "conan\cli\commands\export.py", line 51, in export
  File "conan\api\subapi\export.py", line 18, in export
  File "conans\client\cmd\export.py", line 47, in cmd_export
  File "conans\client\cmd\export.py", line 140, in export_source
  File "conan\tools\files\copy_pattern.py", line 30, in copy
AssertionError

ERROR:

```

**Comment by memsharded** [comment]
Hi @RobertHabrich 

Thanks for the feedback.

Yes, we already did that! :) It has been merged for Conan 2.2, it will be in the next release soon.
